python/test/unit/language/test_core.py

```diff
@@ -6002,6 +6002,37 @@ def kernel(Semaphore, Out, total: tl.constexpr):
     assert out.item() >= 0
 
 
+def test_constexpr_flattens():
+    assert tl.constexpr(tl.constexpr(5)) == tl.constexpr(5)
+    assert tl.constexpr(tl.constexpr(tl.constexpr(5))) == tl.constexpr(5)
+
+
+def test_constexpr_type():
+    from triton.language.core import constexpr_type
+    assert tl.constexpr(5).type != tl.float32
+    assert tl.constexpr(5).type == constexpr_type(5)
+    assert tl.constexpr(tl.float32).type == constexpr_type(tl.float32)
+
+
+@pytest.mark.parametrize("literal", [10, 32.1, (5, 6, 7)])
+def test_constexpr_assignment(literal):
+    from triton.language.core import constexpr_type
+
+    @triton.jit
+    def kernel(input_literal: tl.constexpr):
+        patched_literal: tl.constexpr = GENERATE_TEST_HERE
+        # Sanity checks
+        tl.static_assert(patched_literal.type == constexpr_type)
+        tl.static_assert(input_literal.type == constexpr_type)
+
+        assigned_literal = input_literal
+        tl.static_assert(assigned_literal.type == constexpr_type)
+        tl.static_assert(assigned_literal == patched_literal)
```

Comment:
No, the semantics shouldn't be changed here. Otherwise it will break simple code like
```
var = 0
while test:
    var = var + 1
```

Reply:
Doc updates like this would be out of scope for this PR. If I feel compelled I'll take a stab in another PR.